### PR TITLE
User is not found when is searched by name and one of surnames #7510

### LINKED ISSUE #7510: User is not found when is searched by name and one of surnames
### Premise
User backend in LDAP

### Steps to reproduce
1. Upload a file
2. Click on sharing icon
3. Search for a user, using name and first of surnames. For example: User is "Roger Luis Padilha de Souza". Then, search for "Roger Luis". User will be found.
3. After, search for same user, using name and one for surnames, except the first of them. For example: User is "Roger Luis Padilha de Souza". Then, search for "Roger Padilha"
4. User won't be found

### Expected behaviour
User must be found using name and one of surnames

### Actual behaviour
User is found only if search uses name and first of surnames

### Constraint
Behaviour must be defined by configuration (config.php). Default must be actual behaviour. Expected behaviour must be enabled by key **user_ldap.enable_medial_search** equals to **true**.

### PR DESCRIPTION
**Pull request by fgsl**
Fixes #7510

### CODE REVIEW

**Comment by MorrisJobke** [comment]
cc @nextcloud/ldap

**Comment by MorrisJobke** [comment]
Copying my comment from #7510:

I would still like to have this properly documented instead of adding another hidden feature that adds more code. You could simply search for "*eyer" to find "Meyer" aswell as "Beyer" in the search field.

File: apps/user_ldap/lib/Access.php [hunk 1]
```diff
@@ -1467,12 +1467,17 @@ private function prepareSearchTerm($term) {
 		$config = \OC::$server->getConfig();
 
 		$allowEnum = $config->getAppValue('core', 'shareapi_allow_share_dialog_user_enumeration', 'yes');
+		$allowMedialSearches = $config->getSystemValue('user_ldap.enable_medial_search', false);
```
Comment (blizzz) [inline apps/user_ldap/lib/Access.php]: should not be a system value, but applications specific. And per configuration. Add it to `OCA\User_LDAP\Configuration` and you can read it from `$this->connection` in the Access class.

**Comment by blizzz** [comment]
> I would still like to have this properly documented instead of adding another hidden feature that adds more code. You could simply search for "*eyer" to find "Meyer" aswell as "Beyer" in the search field.

I am not vetoing this… I'd be OK but not super happy either. @fgsl convince us?

**Comment by fgsl** [comment]
>     I would still like to have this properly documented instead of adding another hidden feature that adds more code. You could simply search for "*eyer" to find "Meyer" aswell as "Beyer" in the search field.
> 
> I am not vetoing this… I'd be OK but not super happy either. @fgsl convince us?
> 
@blizzz and @MorrisJobke, I understand the example, but it is not the type of search intended with this issue. Imagine that you have 10 thousands of users, and among them "Meyer" and "Beyer" are very common, so when you find by "*eyer" the result is greater than hundred. But you know that you want to find a "Meyer" that has a "Gunther" in some part of his complete name. And you don't know or remember the complete name. And the complete name is something like Otto Meyer Manfred Gunther Wagner. You types "Meyer Gunther" and there is no result. You only have results for isolated names or for names exactly in order.

File: apps/user_ldap/lib/Access.php (hunk 1 above)
Comment (fgsl) [reply apps/user_ldap/lib/Access.php]: Ok, @blizzz, I will change and submit again.

**Comment by MorrisJobke** [comment]
Then you can also search for "*Meyer Gunther*" and would accomplish the same as what this PR does.

**Comment by fgsl** [comment]
> Then  you can also search for *Meyer Gunther* and would accomplish the same as what this PR does.

@MorrisJobke, I tried to follow this guideline, but using example of [7510](#7510#issuecomment-351737141), "Roger Luis Padilha de Souza". I tried to search for "\*Roger Padilha\*". Nothing comes.

**Comment by MorrisJobke** [comment]
> @MorrisJobke, I tried to follow this guideline, but using example of 7510, "Roger Luis Padilha de Souza". I tried to search for "*Roger Padilha*". Nothing comes.

But this also does not work with the patch you supplied, because it does exactly that. Or am I wrong @blizzz ?

**Comment by blizzz** [comment]
> "*Roger Padilha*"

No asterisk at the end, that gets escaped and is appended by default. Only place it on the front.

**Comment by blizzz** [comment]
> > @MorrisJobke, I tried to follow this guideline, but using example of 7510, "Roger Luis Padilha de Souza". I tried to search for "*Roger Padilha*". Nothing comes.

> But this also does not work with the patch you supplied, because it does exactly that. Or am I wrong @blizzz ?

Correct.  To have this working you have to have each name in an attribute of itself or in a multi-value attribute that you include in the user search attributes.

**Comment by fgsl** [comment]
> Correct. To have this working you have to have each name in an attribute of itself or in a multi-value attribute that you include in the user search attributes.

This change works for our case because we have he complete name of user in one only attribute of LDAP. For searches in more than one attribute of LDAP this change will help because Nextcloud will find user "Blizzz" in a search by "lizzz". If filter has not "*" at beginning, a user never will be found if search doesn't start at first character of name/surname.

**Comment by blizzz** [comment]
> > > @MorrisJobke, I tried to follow this guideline, but using example of 7510, "Roger Luis Padilha de Souza". I tried to search for "*Roger Padilha*". Nothing comes.

> > But this also does not work with the patch you supplied, because it does exactly that. Or am I wrong @blizzz ?

> Correct.  To have this working you have to have each name in an attribute of itself or in a multi-value attribute that you include in the user search attributes.

I need to correct me.  We split up the search terms into words by space delimiter. So it would work. By heart I am not sure whether we keep an asterisk in tact that is within the search term. @fgsl please try "*Roger *Padilha".

**Comment by fgsl** [comment]
> I need to correct me. We split up the search terms into words by space delimiter. So it would work. By heart I am not sure whether we keep an asterisk in tact that is within the search term. @fgsl please try "*Roger *Padilha".

@blizzz , i tried "*Roger *Padilha". It didn't work.

**Comment by blizzz** [comment]
@fgsl ok, thanks. The issue is that the escaping the search term happens directly in the user manager method, e.g. first line in `User_LDAP::getUsers()`, and not when analyzing the string. `Access::escapeFilterPart()`  could be made smarter to also break it down into  words. The advantage is that at this point we know when to allow the prepended asterisk (goes as parameter to escapeFilterPart). 

The break down into words happens further down in the stack at `getAdvancedFilterPartForSearch`. It's essentially a simple oneliner,  `$searchWords = explode(' ', trim($search));` before further processing.  It is applied only with some conditions, see `Access::getFilterPartForSearch()`.

With extracting some methods, `Access::escapeFilterPart()`  could become aware of that and keeps the asterisk in front of words unescaped.

**Comment by fgsl** [comment]
Hi, @blizzz. Sorry, but I feel that awareness of  `Access::escapeFilterPart()` would make all more complicated. After read your post, I thought about it, but, if I really understand, we can fall in something like (not good) idea below:

1) Create two static attributes for Access class:

```
private static $escapeFilterPartSearch = null;
private static $escapeFilterPartReplace = null;
```
2) Create two static methods for Access class:

```
private static function getEscapeFilterPartSearch(){
    if (null == self::$escapeFilterPartSearch){
        self::$escapeFilterPartSearch = ($this->connection->ldapMedialSearch, array('\\', '(', ')'), array('*', '\\', '(', ')'));
    }
    return self::$escapeFilterPartSearch;
}
private static function getEscapeFilterPartReplace(){
    if (null == self::$escapeFilterPartReplace){
        self::$escapeFilterPartReplace = ($this->connection->ldapMedialSearch, array('\\\\', '\\(', '\\)'), array('\\\\', '\\(', '\\)'),);
    }
    return self::$escapeFilterPartReplace;
}
```
3) Change three last lines of Access::escapeFilterPart() for this:

`		return $asterisk . str_replace(self::getEscapeFilterPartSearch(), self::getEscapeFilterPartReplace(), $input);
`

**Comment by MorrisJobke** [comment]
@fgsl You merged master into your branch ... maybe a git rebase onto of current master is a lot better to keep the PR clean. See http://morrisjobke.de/2015/12/03/How-to-do-a-git-rebase/ for some steps to do so, otherwise reach out here so we could guide you :)

**Comment by fgsl** [comment]
> @fgsl You merged master into your branch ... maybe a git rebase onto of current master is a lot better to keep the PR clean. See http://morrisjobke.de/2015/12/03/How-to-do-a-git-rebase/ for some steps to do so, otherwise reach out here so we could guide you :)

Thank you and I'm sorry, @MorrisJobke. I will follow guidelines.

**Comment by fgsl** [comment]
@MorrisJobke and @blizzz , I made a mistake and had to submit a new pull request: https://github.com/nextcloud/server/pull/8760.

**Comment by MorrisJobke** [comment]
> @MorrisJobke and @blizzz , I made a mistake and had to submit a new pull request: #8760.

Don't worry :)

**Pull request closed without merging**